### PR TITLE
PWX-33339: Stashstrategy integration tests for mongodb and kafka operator.

### PR DESCRIPTION
**Pull request by diptiranjanpx**
Signed-Off-By: Diptiranjan

**What type of PR is this?**
>integration-test

**What this PR does / why we need it**:
Tests for stash strategy feature in migration.

**Does this PR change a user-facing CRD or CLI?**:

**Is a release note needed?**:

**Does this change need to be cherry-picked to a release branch?**:

Successful run - https://jenkins.pwx.dev.purestorage.com/job/Users/job/dipti/job/diptistorkcustom/5/

### CODE REVIEW

File: test/integration_test/migration_features_test.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+package integrationtest
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	storkapi "github.com/libopenstorage/stork/pkg/apis/stork/v1alpha1"
```
Comment (olavangad-px) [inline test/integration_test/migration_features_test.go:12]: nit.. duplicate imports storkapi and storkv1

Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go:12]: removed.

File: test/integration_test/migration_features_test.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+			nil,
+			nil,
+			nil,
+			false,
+			resourcecollector.Options{},
+		)
+		if err != nil {
+			return validated, fmt.Errorf("error fetching objects while validating CR: %w", err)
+		}
+		logrus.Infof("number of resources for app %s in namespace %s is %d", appName, namespace, len(objects.Items))
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:228]: We should log GVK here for the resource. Wdyt?

File: test/integration_test/migration_features_test.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+		}
+	}
+	return nil
+}
+
+func validateCR(appName string, namespace string, kubeConfig string) (bool, error) {
+	validated := true
+	operatorCRMap := getSupportedOperatorCRMapping()
+	if _, ok := operatorCRMap[appName]; !ok {
+		return validated, fmt.Errorf("app %s is not currently supported in test framework", appName)
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:198]: nit.. We are returning err and handling in caller. But returning validated = true is a bit misleading. We should return false in all error scenarios.

File: test/integration_test/migration_features_test.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	// validate no stashed cm is present
+	stashedCMCount := getStashedConfigMapCount(t, appData.Ns)
+	expectedStashedCMCount := 0
+	require.Equal(t, expectedStashedCMCount, stashedCMCount, fmt.Sprintf("expected stashed configmap count %d got %d", expectedStashedCMCount, stashedCMCount))
+
+	podsMigrated, err := core.Instance().GetPods(appData.Ns, nil)
+	require.NoError(t, err, "Error getting migrated pods")
+
+	podsMigratedLen := len(podsMigrated.Items)
+	require.Equal(t, podsCreatedLen, podsMigratedLen, "Pods migration failed as len of pods found on source doesnt match with pods found on destination")
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:313]: nit..doesn't

File: test/integration_test/migration_features_test.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+	logrus.Infof("Delete Destination and Source Ns")
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:315]: nit.. destination and source

File: test/integration_test/migration_features_test.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+	require.NoError(t, err, "Error setting source kubeconfig")
+
+	err = asyncdr.DeleteAndWaitForMigrationDeletion(mig.Name, mig.Namespace)
+	require.NoError(t, err, "Error deleting migration")
+
+	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
+	err = core.Instance().DeleteNamespace(appData.Ns)
+	if err != nil {
+		logrus.Errorf("Error deleting namespace %s: %v\n", appData.Ns, err)
+	}
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:330]: Same concern as above.

**Review by pp511 — COMMENTED** [review]
Mostly looks good. Just nitpicking and some questions.

File: test/integration_test/common_test.go [hunk 7]
```diff
@@ -957,6 +957,124 @@ func scheduleUnidirectionalClusterPair(cpName, cpNamespace, projectMappings stri
...
+
+		err = dumpKubeConfigPath(destConfig, destKubeconfigPath)
+		if err != nil {
+			return fmt.Errorf("unable to dump remote config while setting destination config during activate migrations: %v", err)
+		}
+		cmdArgs = append(cmdArgs, []string{"--kubeconfig", destKubeconfigPath}...)
+	}
+
+	cmd.SetArgs(cmdArgs)
+	logrus.Infof("activating apps in namespace with command: %v", cmdArgs)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/common_test.go]: Done

File: test/integration_test/common_test.go [hunk 8]
```diff
@@ -957,6 +957,124 @@ func scheduleUnidirectionalClusterPair(cpName, cpNamespace, projectMappings stri
...
+		}()
+		err = dumpKubeConfigPath(destConfig, destKubeconfigPath)
+		if err != nil {
+			return fmt.Errorf("unable to dump remote config while setting destination config during deactivate migrations: %v", err)
+		}
+		cmdArgs = append(cmdArgs, []string{"--kubeconfig", destKubeconfigPath}...)
+	}
+
+	cmd.SetArgs(cmdArgs)
+	logrus.Infof("deactivating apps in namespace with command: %v", cmdArgs)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/common_test.go]: Done

File: test/integration_test/migration_features_test.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+	require.NoError(t, err, fmt.Sprintf("Error validating CR for app %s in namespace %s", appName, appData.Ns))
+	require.True(t, validated, fmt.Sprintf("CR for app %s in namespace %s should have been present", appName, appData.Ns))
+
+	podsMigrated, err := core.Instance().GetPods(appData.Ns, nil)
+	require.NoError(t, err, "Error getting migrated pods")
+
+	podsMigratedLen := len(podsMigrated.Items)
+	require.Equal(t, podsCreatedLen, podsMigratedLen, "Pods migration failed as len of pods found on source doesnt match with pods found on destination")
+
+	logrus.Infof("Delete Destination and Source Ns")
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+
+	err = asyncdr.DeleteAndWaitForMigrationDeletion(mig.Name, mig.Namespace)
+	require.NoError(t, err, "Error deleting migration")
+
+	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
+	err = core.Instance().DeleteNamespace(appData.Ns)
+	if err != nil {
+		logrus.Errorf("Error deleting namespace %s: %v\n", appData.Ns, err)
+	}
+	logrus.Infof("Test %s ended", t.Name())
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go:166]: Done

File: test/integration_test/migration_features_test.go (hunk 4 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go (hunk 5 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go (hunk 6 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go:332]: > Same concern as above.

Done

File: test/integration_test/migration_features_test.go (hunk 3 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Addressed.

File: test/integration_test/migration_features_test.go [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+	// validate that CR is present
+	validated, err = validateCR(appName, appData.Ns, destClusterConfigPath)
+	require.NoError(t, err, fmt.Sprintf("Error validating CR for app %s in namespace %s", appName, appData.Ns))
+	require.True(t, validated, fmt.Sprintf("CR for app %s in namespace %s should have been present", appName, appData.Ns))
+
+	podsMigrated, err := core.Instance().GetPods(appData.Ns, nil)
+	require.NoError(t, err, "Error getting migrated pods")
+
+	podsMigratedLen := len(podsMigrated.Items)
+	require.Equal(t, podsCreatedLen, podsMigratedLen, "Pods migration failed as len of pods found on source doesnt match with pods found on destination")
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+
+	migrationStashStrategy(t, appNameKafka, appPathKafka)
+	// If we are here then the test has passed
+	testResult = testResultPass
+	logrus.Infof("Test status at end of %s test: %s", t.Name(), testResult)
+}
+
+func migrationStashStrategy(t *testing.T, appName string, appPath string) {
+	err := setSourceKubeConfig()
+	require.NoError(t, err, "failed to set kubeconfig to source cluster: %w", err)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Made it %v everywhere, as in logrus.Errorf is not able to wrap %w

File: test/integration_test/common_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply test/integration_test/common_test.go]: Done

File: test/integration_test/migration_features_test.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+	err = setSourceKubeConfig()
+	require.NoError(t, err, "Error setting source kubeconfig")
+
+	err = asyncdr.DeleteAndWaitForMigrationDeletion(mig.Name, mig.Namespace)
+	require.NoError(t, err, "Error deleting migration")
+
+	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
+	err = core.Instance().DeleteNamespace(appData.Ns)
+	if err != nil {
+		logrus.Errorf("Error deleting namespace %s: %v\n", appData.Ns, err)
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:331]: nit..Error deleting namespace on source

File: test/integration_test/migration_features_test.go [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+	require.NoError(t, err, "Error getting migrated pods")
+
+	podsMigratedLen := len(podsMigrated.Items)
+	require.Equal(t, podsCreatedLen, podsMigratedLen, "Pods migration failed as len of pods found on source doesn't match with pods found on destination")
+
+	logrus.Infof("Delete destination and source namespaces")
+	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
+	err = core.Instance().DeleteNamespace(appData.Ns)
+	if err != nil {
+		logrus.Errorf("Error deleting namespace %s: %v\n", appData.Ns, err)
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:320]: nit..Error deleting namespace on destination

File: test/integration_test/migration_features_test.go (hunk 2 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go:229]: the check expects the number of CRs to be present is 1. 
In most cases, if migration has not applied the CR or it has failed, the count will be 0 (no need to print GVK in that case). And for count 1 , its validated. 

Anyway the app's related CR info is already predefined in the test, so debugging is not difficult also.
So I think, we can skip printing GVK for each object in this line.

File: test/integration_test/migration_features_test.go [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:153]: Error deleting namespace on destination

File: test/integration_test/migration_features_test.go [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+	err = setSourceKubeConfig()
+	require.NoError(t, err, "Error setting source kubeconfig")
+
+	err = asyncdr.DeleteAndWaitForMigrationDeletion(mig.Name, mig.Namespace)
+	require.NoError(t, err, "Error deleting migration")
+
+	asyncdr.DeleteCRAndUninstallCRD(appData.OperatorName, appPath, appData.Ns)
+	err = core.Instance().DeleteNamespace(appData.Ns)
+	if err != nil {
+		logrus.Errorf("Error deleting namespace %s: %v\n", appData.Ns, err)
```
Comment (pp511) [inline test/integration_test/migration_features_test.go:164]: Error deleting namespace on source

**Review by pp511 — APPROVED** [review]
Just few more nits for adding more logs during CRD deletion.

File: test/integration_test/migration_features_test.go (hunk 13 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go (hunk 15 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go (hunk 16 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

File: test/integration_test/migration_features_test.go (hunk 14 above)
Comment (diptiranjanpx) [reply test/integration_test/migration_features_test.go]: Done

**Review by gejain — APPROVED** [review]
LGTM

**Comment by diptiranjanpx** [comment]
Updated test result: https://jenkins.pwx.dev.purestorage.com/job/Users/job/dipti/job/diptistorkcustom/9/

**Pull request merged**